feat(RHINENG-19064) Add script to delete hosts by IDs from s3 bucket

kruai opened the pull request:
# Overview

This PR is being created to address [RHINENG-19064](https://issues.redhat.com/browse/RHINENG-19064).
Adds a script that deletes hosts specified by the `hbi-delete-host-ids-s3` bucket's contents. For each ID in the bucket, it attempts to find the host, and if the host has only one reporter, deletes it (if it's not a dry run). Some notes:

- No hosts will be deleted if the script is running in dry-run mode, which is the default.
- The script keeps track of the number of hosts that are deleted, not found, and not deleted due to having multiple reporters. It logs this data at the end at INFO level.
- By default, this script processes hosts in batches of 500 to reduce DB load. For 4.5 million hosts, that will be ~9K batches. This is configurable via the `DELETE_HOSTS_S3_BATCH_SIZE` env var.
- This now streams data from s3 rather than loading in the whole bucket at once.

## PR Checklist

- [x] Keep PR title short, ideally under 72 characters
- [x] Descriptive comments provided in complex code blocks
- [ ] Include raw query examples in the PR description, if adding/modifying SQL query
- [x] Tests: validate optimal/expected output
- [x] Tests: validate exceptions and failure scenarios
- [x] Tests: edge cases
- [x] Recovers or fails gracefully during potential resource outages (e.g. DB, Kafka)
- [x] Uses [type hinting](https://docs.python.org/3/library/typing.html), if convenient
- [ ] Documentation, if this PR changes the way other services interact with host inventory
- [ ] Links to related PRs

## Secure Coding Practices Documentation Reference

You can find documentation on this checklist [here](https://github.com/RedHatInsights/secure-coding-checklist).

## Secure Coding Checklist

- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

## Summary by Sourcery

Introduce a new batch-driven host deletion job that reads host IDs from a CSV file in an S3 bucket and removes hosts with only one reporter, supporting dry-run and logging summaries, and integrate it into the deployment manifest.

New Features:
- Add delete_hosts_s3.py script for fetching host IDs from S3 and deleting eligible hosts in configurable batches with optional dry-run mode
- Add a delete-hosts-s3 Kubernetes job to the clowdapp deployment manifest with required environment variables, secrets, and resource settings

Enhancements:
- Expose DRY_RUN flag and introduce DELETE_HOSTS_S3_AWS_SECRET parameter for the new job
- Rename S3_AWS_SECRET to KESSEL_S3_AWS_SECRET for the Kessel S3 bucket secret reference

Tests:
- Add unit tests for delete_hosts_s3.run covering batching logic, dry-run behavior, exception handling, and resource cleanup

kruai commented on delete_hosts_s3.py:
```diff
@@ -0,0 +1,139 @@
+    with application.app.app_context():
+        try:
+            logger.info(f"Running host_delete_s3 with batch size {BATCH_SIZE}")
+            s3_client = get_s3_client(config)
```
The s3 client does support `close()`.

kruai commented on delete_hosts_s3.py:
```diff
@@ -0,0 +1,139 @@
+                not_deleted_count += 1
+            else:
+                # Delete the host if it's not in dry-run mode
+                if not config.dry_run:
```
I've updated the variable to be a boolean.

ezr-ondrej reviewed (dismissed):
Thanks for addressing this, it's pretty annoying to have stale hosts :)
Just two code smells I've noticed, apart from it looks good to go to me 👍

kruai commented on delete_hosts_s3.py:
```diff
@@ -0,0 +1,139 @@
+        config (Config): Application configuration object.
+        logger (Logger): Logger for logging information.
+    """
+    global deleted_count, not_deleted_count, not_found_count
```
I'm declaring these to be global variables here so it doesn't create new local vars when it assigns them in the function. It's a little weird, but I double-checked this usage :)

fstavela commented on delete_hosts_s3.py:
```diff
@@ -0,0 +1,139 @@
+            else:
+                # Delete the host if it's not in dry-run mode
+                if not config.dry_run:
+                    session.delete(host)
```
Shouldn't we use `delete_hosts` from `lib.host_delete`? My main concern is that some of the hosts may be in groups and if we just delete them from the hosts table, it could cause issues. The other reason would be that using `delete_hosts` function would also send events and notifications. To make it more efficient, we could just save the host_ids of the hosts to be deleted and then use `delete_hosts` on all of them when we process the full batch

kruai commented on delete_hosts_s3.py:
```diff
@@ -0,0 +1,139 @@
+                deleted_count += 1
+
+        # Commit deletions if not in dry-run mode
+        if not config.dry_run:
+            session.commit()
```
Yeah, it's implicit in `session_guard`. Sourcery was (locally) suggesting I add an explicit commit here, but I can remove it.

fstavela commented on app/config.py:
```diff
@@ -343,6 +343,7 @@ def __init__(self, runtime_environment):
         self.s3_access_key_id = os.getenv("S3_AWS_ACCESS_KEY_ID")
         self.s3_secret_access_key = os.getenv("S3_AWS_SECRET_ACCESS_KEY")
         self.s3_bucket = os.getenv("S3_AWS_BUCKET")
+        self.dry_run = os.getenv("DRY_RUN", "true").lower() == "true"
```
Aaah, that makes sense, I didn't notice that the `DRY_RUN` is mapped to a different variable in the config.

kruai commented on delete_hosts_s3.py:
```diff
@@ -0,0 +1,161 @@
+    """
+    global deleted_count, not_deleted_count, not_found_count
+    with session_guard(session):
+        logger.info(f"Processing batch of {len(batch)} host IDs...")
```
Updated this log, as well as the docstrings.

kruai commented:
> You could change the "host IDs" in the docstrings and logs to "subscription manager IDs", but it's not a blocker.
> 
> I think the script will need the kafka variables in the config now, but otherwise the PR looks good!

@fstavela You're right. I added the Kafka env vars, and updated the logs and docstrings.

fstavela reviewed (approved):
LGTM, thank you!
